File: problems/watchlater/submissions/accepted/watchlater-zf.cpp
#include <iostream>
#include <cstring>
#include <set>

using namespace std;

char watch[401];
int dp[1<<20], toint[256];
bool seen[26];
int n, k;

int solve(int rem) {
  if (!rem) return 0;

  if (dp[rem] == -1) {
    int clicks[20] = {0};
    int prev = -1;

    for (int i = 0; watch[i]; ++i) {
      int c = toint[watch[i]];
      if (rem&(1<<c)) {
        if (prev != c) ++clicks[c];
        prev = c;
      }
    }

    dp[rem] = 1000;
    for (int i = 0; i < k; ++i)
      if (rem&(1<<i))
        dp[rem] = min(dp[rem], clicks[i] + solve(rem^(1<<i)));
  }
  return dp[rem];
}

int main() {
  cin >> n >> k >> watch;

  memset(dp, -1, sizeof(dp));
  memset(toint, -1, sizeof(toint));

  int x = 0;
  for (int i = 0; watch[i]; ++i)
    if (toint[watch[i]] == -1)
      toint[watch[i]] = x++;

  cout << solve((1<<k)-1) << endl;
  return 0;
}

File: problems/watchlater/submissions/time_limit_exceeded/watchlater-zf.py
#!/usr/bin/python3

def solve(rem):
    if rem == 0: return 0

    if rem not in memo:
        clicks = [0]*20
        prev = -1
        for c in watch:
            if rem&(1<<c) and prev != c:
                clicks[c] += 1
                prev = c
        memo[rem] = min(clicks[c] + solve(rem^(1<<c)) for c in range(k) if rem&(1<<c))

    return memo[rem]

n, k = map(int, input().split())
watch = input()

memo = {}
toint = {}

for c in watch:
    if c not in toint:
        toint[c] = len(toint)

watch = [toint[c] for c in watch]

print(solve((1<<k)-1))
